pkg/kademlia/queue.go

```diff
@@ -4,152 +4,123 @@
 package kademlia
 
 import (
-	"container/heap"
-	"math/big"
+	"math/bits"
+	"sort"
 	"sync"
 
 	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/storj"
 )
 
-// XorQueue is a priority queue where the priority is key XOR distance
-type XorQueue struct {
+// Queue is a limited priority queue with xor distance
+type Queue struct {
 	maxLen int
+	mu     sync.Mutex
+	added  map[storj.NodeID]int
+	items  []queueItem
+}
 
-	mu    sync.Mutex
-	added map[storj.NodeID]int
-	items items
+// An queueItem is something we manage in a priority queue.
+type queueItem struct {
+	node     *pb.Node
+	priority storj.NodeID
+	// TODO: switch to using pb.NodeAddress to avoid pointer to *pb.Node
 }
 
-// NewXorQueue returns a items with priority based on XOR from targetBytes
-func NewXorQueue(size int) *XorQueue {
-	return &XorQueue{
-		items:  make(items, 0, size),
+// xorNodeID returns the xor of each byte in NodeID
+func xorNodeID(a, b storj.NodeID) storj.NodeID {
+	r := storj.NodeID{}
+	for i, av := range a {
+		r[i] = av ^ b[i]
+	}
+	return r
+}
+
+// reverseNodeID reverses NodeID bit representation
+func reverseNodeID(a storj.NodeID) storj.NodeID {
+	r := storj.NodeID{}
+	for i, v := range a {
+		r[len(a)-i-1] = bits.Reverse8(v)
+	}
+	return r
+}
+
+// NewQueue returns a items with priority based on XOR from targetBytes
+func NewQueue(size int) *Queue {
+	return &Queue{
 		added:  make(map[storj.NodeID]int),
 		maxLen: size,
 	}
 }
 
-// Insert adds Nodes onto the queue
-func (x *XorQueue) Insert(target storj.NodeID, nodes []*pb.Node) {
-	x.mu.Lock()
-	defer x.mu.Unlock()
+// Insert adds nodes into the queue.
+func (queue *Queue) Insert(target storj.NodeID, nodes ...*pb.Node) {
+	queue.mu.Lock()
+	defer queue.mu.Unlock()
 
 	unique := nodes[:0]
 	for _, node := range nodes {
 		nodeID := node.Id
-		if _, added := x.added[nodeID]; !added {
-			x.added[nodeID]++
+		if _, added := queue.added[nodeID]; !added {
+			queue.added[nodeID]++
 			unique = append(unique, node)
 		}
 	}
 
-	x.insert(target, unique)
+	queue.insert(target, unique...)
 }
 
-// Reinsert adds a Nodes onto the queue if it's been added >= limit times previously
-func (x *XorQueue) Reinsert(target storj.NodeID, node *pb.Node, limit int) bool {
-	x.mu.Lock()
-	defer x.mu.Unlock()
+// Reinsert adds a Nodes into the queue, only if it's has been added less than limit times.
+func (queue *Queue) Reinsert(target storj.NodeID, node *pb.Node, limit int) bool {
```

Comment:
The use of this function in peer_discovery.go doesn't really match its description; could limit be unneeded?

Reply:
I'm not following what you mean by "doesn't match its description"?